File: src/main/java/club/deneb/client/client/FriendManager.java
package club.deneb.client.client;

import club.deneb.client.utils.ChatUtil;
import club.deneb.client.utils.clazz.Friend;
import net.minecraft.entity.Entity;

import java.util.ArrayList;
import java.util.List;

/**
 * Author B_312 on 01/15/2021
 */
public class FriendManager {

    public static FriendManager INSTANCE;

    public FriendManager(){
        INSTANCE = this;
    }

    public ArrayList<Friend> friends = new ArrayList<>();

    public boolean checkExist(String name){
        for(Friend friend : friends){
            if(friend.name.equalsIgnoreCase(name)) return true;
        }
        return false;
    }

    public Friend getFriendByName(String name){
        for(Friend friend : friends){
            if(friend.name.equalsIgnoreCase(name)) return friend;
        }
        return null;
    }

    public static void addFriend(String name) {
        if(isNull()) return;
        if(!INSTANCE.checkExist(name)) {
            INSTANCE.friends.add(new Friend(name,true));
        } else {
            INSTANCE.getFriendByName(name).isFriend = true;
        }
        ChatUtil.sendNoSpamMessage(ChatUtil.SECTIONSIGN + "b" + name + ChatUtil.SECTIONSIGN + "r has been friended.");
    }

    public static void removeFriend(String name) {
        if(isNull()) return;
        if(INSTANCE.checkExist(name)) {
            INSTANCE.getFriendByName(name).isFriend = false;
            ChatUtil.sendNoSpamMessage(ChatUtil.SECTIONSIGN + "b" + name + ChatUtil.SECTIONSIGN + "r has been unfriended.");
        } else {
            ChatUtil.sendNoSpamMessage(ChatUtil.SECTIONSIGN + "b" + name + ChatUtil.SECTIONSIGN + "r is not our friend!");
        }
    }

    public static boolean isNull(){
        return INSTANCE == null;
    }

    public static boolean isFriend(String name) {
        if(isNull()) return false;
        if(!INSTANCE.checkExist(name)) return false;
        return INSTANCE.getFriendByName(name).isFriend;
    }

    public static boolean isFriend(Entity entity){
        return isFriend(entity.getName());
    }

    public static List<Friend> getFriendList(){
        if(isNull()) return new ArrayList<>();
        return INSTANCE.friends;
    }

    public static List<String> getFriendStringList(){
        if(isNull()) return new ArrayList<>();
        List<String> stringList = new ArrayList<>();
        getFriendList().forEach(f -> stringList.add(f.name));
        return stringList;
    }

}

File: src/main/java/club/deneb/client/client/FriendManager.kt
package club.deneb.client.client

import club.deneb.client.utils.ChatUtil
import net.minecraft.entity.Entity

/**
 * Author B_312 on 01/15/2021
 * Updated by B_312 on 04/24/2021
 */
object FriendManager {

    var friends = mutableMapOf<String,Boolean>()

    @JvmStatic
    fun addFriend(name: String) {
        when{
            isFriend(name) -> {
                ChatUtil.sendNoSpamMessage(ChatUtil.SECTIONSIGN.toString() + "b" + name + ChatUtil.SECTIONSIGN + "r has already been our friend.")
            }
            friends[name] != null && friends[name] == false -> {
                ChatUtil.sendNoSpamMessage(ChatUtil.SECTIONSIGN.toString() + "b" + name + ChatUtil.SECTIONSIGN + "r is our friend again.")
            }
            else -> {
                ChatUtil.sendNoSpamMessage(ChatUtil.SECTIONSIGN.toString() + "b" + name + ChatUtil.SECTIONSIGN + "r is our friend now.")
            }
        }
        friends[name] = true //Kotlin strong
    }

    @JvmStatic
    fun removeFriend(name: String) {
        when{
            isFriend(name) -> {
                ChatUtil.sendNoSpamMessage(ChatUtil.SECTIONSIGN.toString() + "b" + name + ChatUtil.SECTIONSIGN + "r is no longer our friend.")
            }
            friends[name] != null && friends[name] == false -> {
                ChatUtil.sendNoSpamMessage(ChatUtil.SECTIONSIGN.toString() + "b" + name + ChatUtil.SECTIONSIGN + "r is not our friend.")
            }
            else -> {
                ChatUtil.sendNoSpamMessage(ChatUtil.SECTIONSIGN.toString() + "b" + name + ChatUtil.SECTIONSIGN + "r hasn't been our friend yet.")
            }
        }
        friends[name] = false //Kotlin strong
    }

    @JvmStatic
    fun isFriend(name: String): Boolean {
        return friends[name] != null && friends[name] == true
    }

    @JvmStatic
    fun isFriend(entity: Entity): Boolean {
        return isFriend(entity.name)
    }

    fun friendList(): List<String> {
        return friends.keys.toList()
    }

}